DEIMOS force extract

debora-pe opened the pull request:
The goal of this PR is to allow PypeIt to force an extraction of DEIMOS undetected objects at the location expected from the slitmask design. A few more things are also done.

Related:
- Improved `slittrace.assign_maskinfo`
- Force extraction of non detected objects using the function `slittrace.mask_add_missing_obj`
- Added new parameters to `SlitMaskPar`
- The forced extractions are traced by a flag in the SpecObj datamodel.
- `pypeit_show_2dspec` shows the traces of forced extraction with a different color (yellow) 
- the txt file and `pypeit_show_1dspec` report the flag for the forced extraction
-  added report
- updated docs
- update version of SpecObj and SlitTrace datamodels

Non related:
- `use_maskdesign` is turned off for LVM masks 
- a new parameter "optimal_fwhm" is added in `ExtractionPar ` to allow the user to set their preferred fwhm

Changes to the Dev Suite to accompany this PR are in https://github.com/pypeit/PypeIt-development-suite/pull/145

<img width="2066" alt="missing_objects_ginga" src="https://user-images.githubusercontent.com/16786072/113493241-7391c080-9492-11eb-9364-cd7254d593b9.png">

profxj committed:
starting up

profxj committed:
working version?

profxj committed:
wip

profxj committed:
ran through!

profxj committed:
fussin

debora-pe committed:
revamp assign_maskinfo and updates to mask_add_missing_obj

debora-pe committed:
Merge branch 'develop' into deimos_force_extract

debora-pe committed:
print info on force_extract

debora-pe committed:
testing no median_offset

debora-pe committed:
minor things

debora-pe committed:
LVM,default parset, update specobj datamodel

debora-pe committed:
final

debora-pe committed:
Merge branch 'develop' into deimos_force_extract

debora-pe committed:
add unit test

debora-pe committed:
update files for unit tests

debora-pe committed:
report

debora-pe committed:
fix some docs

debora-pe committed:
small fixes

jhennawi commented on pypeit/core/extract.py:
```diff
@@ -1652,6 +1658,7 @@ def objfind(image, thismask, slit_left, slit_righ, inmask=None, fwhm=3.0, maxdev
     qobj = np.zeros_like(xtmp)
     for iobj in range(nobj):
         # TODO -- This parameter may not be used anywhere
+        #   It is in skysub.py at line 674
```
Please remove this comment. It is definitely used.

profxj replied:
and remote the TODO too.

debora-pe replied:
Done

debora-pe commented:
Dev Suite passed and cooked created (but not uploaded yet)

```
--- PYPEIT DEVELOPMENT SUITE FAILED 1/100 TESTS (Masters ignored) ---
Failed tests:
    gemini_gnirs/32_SB_SXD pypeit_coadd_1dspec
Skipped tests:
    gemini_gnirs/32_SB_SXD pypeit_coadd_2dspec
    gemini_gnirs/32_SB_SXD pypeit_tellfit
Testing Started at 2021-04-02T20:43:20.002983
Testing Completed at 2021-04-03T03:04:40.311182
Total Time: 6:21:20.308199
(pypeit) dpelliccia@ukulele2~/PypeIt/PypeIt-development-suite$ ./pypeit_test gemini_gnirs -s 32_SB_SXD
Running tests on the following instruments:
    gemini_gnirs

Reducing data from gemini_gnirs for the following setups:
    32_SB_SXD

 0 passed/ 0 failed/ 0 skipped STARTED gemini_gnirs/32_SB_SXD pypeit (without masters)
 1 passed/ 0 failed/ 0 skipped PASSED  gemini_gnirs/32_SB_SXD pypeit (without masters)
 1 passed/ 0 failed/ 0 skipped STARTED gemini_gnirs/32_SB_SXD pypeit_sensfunc
 2 passed/ 0 failed/ 0 skipped PASSED  gemini_gnirs/32_SB_SXD pypeit_sensfunc
 2 passed/ 0 failed/ 0 skipped STARTED gemini_gnirs/32_SB_SXD pypeit_flux_setup
 3 passed/ 0 failed/ 0 skipped PASSED  gemini_gnirs/32_SB_SXD pypeit_flux_setup
 3 passed/ 0 failed/ 0 skipped STARTED gemini_gnirs/32_SB_SXD pypeit_flux
 4 passed/ 0 failed/ 0 skipped PASSED  gemini_gnirs/32_SB_SXD pypeit_flux
 4 passed/ 0 failed/ 0 skipped STARTED gemini_gnirs/32_SB_SXD pypeit_coadd_1dspec
 5 passed/ 0 failed/ 0 skipped PASSED  gemini_gnirs/32_SB_SXD pypeit_coadd_1dspec
 5 passed/ 0 failed/ 0 skipped STARTED gemini_gnirs/32_SB_SXD pypeit_coadd_2dspec
 6 passed/ 0 failed/ 0 skipped PASSED  gemini_gnirs/32_SB_SXD pypeit_coadd_2dspec
 6 passed/ 0 failed/ 0 skipped STARTED gemini_gnirs/32_SB_SXD pypeit_tellfit
 7 passed/ 0 failed/ 0 skipped PASSED  gemini_gnirs/32_SB_SXD pypeit_tellfit

--- PYPEIT DEVELOPMENT SUITE PASSED 7/7 TESTS (Masters ignored) ---
Testing Started at 2021-04-03T11:23:02.911571
Testing Completed at 2021-04-03T12:31:22.255638
Total Time: 1:08:19.344067
```

and Unit tests passed

`============================== 259 passed in 1018.81s (0:16:58) ===========================================`

jhennawi commented on pypeit/core/extract.py:
```diff
@@ -1081,7 +1082,7 @@ def create_skymask_fwhm(sobjs, thismask, box_pix=None):
 
 
-def objfind(image, thismask, slit_left, slit_righ, inmask=None, fwhm=3.0, maxdev=2.0, ir_redux=False, spec_min_max=None,
+def objfind(image, thismask, slit_left, slit_righ, inmask=None, fwhm=3.0, optimal_fwhm=None, maxdev=2.0, ir_redux=False, spec_min_max=None,
```
Will the users ever be setting this parameter to anything besides the expected seeing? I assume no, since that would require specifying a FWHM for every object (i.e. if they are extended objects). 

The code is trying to automatically determine the FWHM from the flux profile. I can see the point of wanting to have control over this, and hence the optimal_fwhm parameter. But a cleaner implementation would change the name of optimal_fwhm to be something like use_user_fwhm, and have this be a Boolean variable. In that regime, it would then use the fwhm parameter always as the fwhm, rather than trying to determine it automatically. 
My point is that the user should be setting the fwhm parameter to be the seeing, and this is the only plausible value for optimal_fwhm. So if someone wants to use their own fwhm, then they would set the use_user_fwhm =True, and then make sure fwhm is set to the seeing. In principle we could just automate this, and always default to fwhm for low S/N objects. 

Either way this opens up a bit of a can of worms, since the object finding may find multiple objects on a slit, and so if you set this parameter it is going to disable the automated FWHM calculation for all objects. We should probably discuss.

debora-pe replied:
I have followed your suggestions and I have changed `optimal_fwhm` with a boolean `use_user_fwhm`

jhennawi commented on pypeit/core/extract.py:
```diff
@@ -1467,55 +1470,58 @@ def objfind(image, thismask, slit_left, slit_righ, inmask=None, fwhm=3.0, maxdev
-                i2 = ind_righ[0]
-                if i2 == 0:
-                    xrigh = None
+        if optimal_fwhm is not None:
```
Following up on above, my suggested change is

if use_user_fwhm: 
   sobjs[iobj].FWHM = fwhm

debora-pe replied:
Done.

jhennawi commented on pypeit/core/skysub.py:
```diff
@@ -522,6 +522,9 @@ def local_skysub_extract(sciimg, sciivar, tilts, waveimg, global_sky, rn2_img,
             that will be used for local sky subtraction. This maskwidth
             is defined in the obfjind code, but is then updated here as
             the profile fitting improves the fwhm estimates
+        user_fwhm: bool, default = False
```
The documentation is not correct here. What you are doing below is forcing a gaussian profile fit with the fwhm determined by the object finding code. This is fine and probably desirable for some cases, but the variable should be called something else like force_gaussian_extraction. It has nothing to do with the fwhm, since you are using whatever fwhm was determined by object finding. You need to be careful with this, since if this is set for high S/N ratio objects you are going to mask a lot of pixels if the true profile does not look a gaussian, which it won't at high S/N ratio. So you need to also turn of the masking in the local sky subtraction.

profxj replied:
Worth noting the masking issue in the docs.

debora-pe replied:
Done as you suggested.

jhennawi commented on pypeit/core/skysub.py:
```diff
@@ -750,12 +753,20 @@ def local_skysub_extract(sciimg, sciivar, tilts, waveimg, global_sky, rn2_img,
-                        thisfwhm=sobjs[iobj].FWHM, maskwidth=sobjs[iobj].maskwidth,
-                        prof_nsigma=sobjs[iobj].prof_nsigma, sn_gauss=sn_gauss, obj_string=obj_string,
-                        show_profile=show_profile)
+                    if user_fwhm is False:
```
This should be changed to: 

profile_model, trace_new, fwhmfit, med_sn2 = extract.fit_profile(
                             sign*img_minsky[ipix], (modelivar * outmask)[ipix],waveimg[ipix], thismask[ipix], spat_pix[ipix], sobjs[iobj].TRACE_SPAT,
                             wave, sign*flux, fluxivar, inmask = outmask[ipix],
                             thisfwhm=sobjs[iobj].FWHM, maskwidth=sobjs[iobj].maskwidth,
                             prof_nsigma=sobjs[iobj].prof_nsigma, sn_gauss=sn_gauss, gauss=user_fwhm, obj_string=obj_string, show_profile=show_profile)

No need for the if/else. But seem my comment above about changing the name of the variable to indicate what it actually does and the not about triggering lots of masking for high S/N ratio objects. Again, let's discuss.

debora-pe replied:
Done

jhennawi commented on pypeit/par/pypeitpar.py:
```diff
@@ -1584,6 +1585,23 @@ def __init__(self, obj_toler=None, assign_obj=None):
+        descr['force_extract'] = 'Force extraction of undetected objects at the location expected ' \
+                                 'from the slitmask design.'
+
+        defaults['force_fwhm'] = None
```
I don't think you need this extra parameter. The user should always be setting the fwhm parameter of the object finding code to the expected fwhm of the objects which is usually going to be the seeing estimate since you only have a single parameter for all objects. There is no need for two paramters. I think you want to always just use that fwhm if force_extract = True. See my comments above in objfind.

debora-pe replied:
I removed

jhennawi commented on pypeit/par/pypeitpar.py:
```diff
@@ -3505,7 +3523,7 @@ class ExtractionPar(ParSet):
     def __init__(self, boxcar_radius=None, std_prof_nsigma=None, sn_gauss=None,
                  model_full_slit=None, manual=None, skip_optimal=None,
-                 use_2dmodel_mask=None):
+                 use_2dmodel_mask=None, optimal_fwhm=None):
```
This should be changed to a boolean use_user_fwhm, or something appropriate to indicate that automatic fwhm determination is being disabled, as we do for manual objects.

debora-pe replied:
Done

jhennawi commented on pypeit/reduce.py:
```diff
@@ -389,9 +389,20 @@ def run(self, ra=None, dec=None, obstime=None, std_trace=None, show_peaks=False,
+                                                             self.par['calibrations']['slitedges']['det_buffer'],
+                                                             TOLER=self.par['reduce']['slitmask']['obj_toler'])
+            # force extraction of non detected objects
+            if self.par['reduce']['slitmask']['force_extract']:
```
I am not following why you are not using the manual object infrastructure here, and instead have created a separate framework for adding objects. Is there any difference between what you are doing here and adding manual objects? I agree that the current framework is not very clean or easy to understand, but we basically have two ways of manually adding objects to slits, which is not very desirable.

profxj replied:
I disagree.  The `manual` approach is qualitatively different than assigning by mask design.

It would be useful, however, to have a single method in `SpecObj` for vetting the addition of a new object (took me awhile to get all the required bits and pieces).

jhennawi commented on pypeit/reduce.py:
```diff
@@ -994,6 +1005,9 @@ def find_objects_pypeline(self, image, std_trace=None,
             boxcar_rad_skymask = self.par['reduce']['extraction']['boxcar_radius'] / self.get_platescale(None),
         else:
             boxcar_rad_skymask = None
+        # user provided optimal fwhm in pixels
```
See my comment above, this should simply be a boolean that you pass to object finding.

debora-pe replied:
Done

jhennawi commented on pypeit/reduce.py:
```diff
@@ -1097,22 +1116,28 @@ def local_skysub_extract(self, global_sky, sobjs, spat_pix=None, model_noise=Tru
-                    show_profile=show_profile,
-                    use_2dmodel_mask=self.par['reduce']['extraction']['use_2dmodel_mask'],
-                    no_local_sky=self.par['reduce']['skysub']['no_local_sky'])
+                try:
```
Please remove this try/except, looks like something being used to debug.

debora-pe replied:
Done

jhennawi commented on pypeit/scripts/show_1dspec.py:
```diff
@@ -47,6 +47,8 @@ def main(args):
                     sobjs[ii].RA,
                     sobjs[ii].DEC,
                     sobjs[ii].MASKDEF_OBJNAME)
+            if sobjs[ii].FORCE_EXTRACT is not None and sobjs[ii].FORCE_EXTRACT is True:
```
Again, we need to be doing the exact same thing for manual objects, not just deimos manual objects.

profxj replied:
That's not clear.  There is a difference between the code forced an extraction and the user chose to via the PypeIt file.

debora-pe replied:
I added another attribute called `MANUAL_EXTRACT` to SpecObj. This will also show here

jhennawi commented on pypeit/scripts/show_2dspec.py:
```diff
@@ -64,11 +64,15 @@ def show_trace(specobjs, det, viewer, ch):
         else:
             trc_name = obj_id
-        display.show_trace(viewer, ch, trace, trc_name, color='orange') #hdu.name)
+        if force_extr_flag is not None and force_extr_flag is True:
```
I like this, but again consistency with manual objects is desired.

profxj replied:
We might want yet another color then.

debora-pe replied:
I added a flag for manual extraction and it will have a different color in ginga.

jhennawi commented on pypeit/scripts/show_2dspec.py:
```diff
@@ -157,20 +161,26 @@ def main(args):
 
     # SKRESIDS
     if 2 in show_channels:
+        # the block below is repeated because if showing this channel but not channel 1 it will crash
```
I am not following what the purpose of the ignore_extract_mask argument is. This was admittedly added in another PR, but perhaps whomever added it can explain.

profxj replied:
This allows one to view the data without the EXTRACT mask being applied.

jhennawi commented on pypeit/slittrace.py:
```diff
@@ -721,140 +723,296 @@ def slit_spat_pos(left, right, nspat):
         return (left[nspec//2,:] + right[nspec//2,:])/2/nspat
 
-    def assign_maskinfo(self, sobjs, plate_scale, TOLER=1.):
+    def mask_add_missing_obj(self, sobjs, expected_objpos, fwhm, median_off, slits_left, slits_right):
```
This code should replace our current manual object infrastructure, and should not require deimos specific mask design information to run. That can easily be done by checking whether that information exists, but the more daunting task is getting rid of the old manual extraction infrastructure.

profxj replied:
I disagree.  These have very different inputs and control flows.
And, as noted, it is a bit too daunting.

jhennawi commented on pypeit/slittrace.py:
```diff
@@ -721,140 +723,296 @@ def slit_spat_pos(left, right, nspat):
+            thisobj.maskwidth = 4. * fwhm  # matches objfind() in extract.py
+            # Finishing up
+            thisobj.set_name()
+            # Mask info
```
Here for example you could just test whether the the mask design information is available before setting these fields.

jhennawi commented on pypeit/specobj.py:
```diff
@@ -167,6 +167,8 @@ class SpecObj(datamodel.DataContainer):
                  'DEC': dict(otype=float, descr='Declination (J2000) decimal degree'),
                  'MASKDEF_ID': dict(otype=(int, np.integer), descr='Slitmask definition ID'),
                  'MASKDEF_OBJNAME': dict(otype=str, descr='Name of the object from the slitmask definition'),
+                 'FORCE_EXTRACT': dict(otype=bool, descr='Boolean indicating if this is a forced extraction '
```
We need to make sure all manually extracted objects receive a true here.

debora-pe replied:
Added another attribute for MANUAL_EXTRACT

jhennawi commented on pypeit/specobj.py:
```diff
@@ -555,3 +557,25 @@ def to_xspec1d(self, **kwargs):
         # Create
         return xspectrum1d.XSpectrum1D.from_tuple((wave, flux, sig))
 
+    def __repr__(self):
```
Nice!

jhennawi commented on pypeit/specobjs.py:
```diff
@@ -711,6 +712,8 @@ def write_info(self, outfile, pypeline):
             if specobj.RA is not None:
                 objra.append(specobj.RA)
                 objdec.append(specobj.DEC)
+            if specobj.FORCE_EXTRACT is not None:
```
also should be done for manual extract objects.

debora-pe replied:
Done

jhennawi commented on pypeit/specobjs.py:
```diff
@@ -747,6 +750,9 @@ def write_info(self, outfile, pypeline):
             obj_tbl['opt_fwhm'].unit = units.arcsec
             obj_tbl['s2n'] = s2n
             obj_tbl['s2n'].format = '.2f'
+            # is this a forced extraction at the expected position from slitmask design?
```
Also update for manual objects.

debora-pe replied:
Done

jhennawi commented on pypeit/spectrographs/keck_deimos.py:
```diff
@@ -255,6 +256,9 @@ def config_specific_par(self, scifile, inp_par=None):
             par['reduce']['slitmask']['assign_obj'] = True
+            # force extraction of undetected objects
+            par['reduce']['slitmask']['force_extract'] = True
+            par['reduce']['slitmask']['force_fwhm'] = 0.5  # arcsec
```
force_fwhm is not necessary as discussed above.

debora-pe replied:
Removed

jhennawi reviewed (changes requested):
Thanks for the very nice work on this PR @deborape. My main substantive feedback is that as of this PR, we will now have two different infrastructures for adding manual objects to object finding and PypeIt reductions. One is the old manual infrastructure, the other is now this deimos specific code. This seems rather undesirable to me. Better would be to replace the old manual infrastructure with your current implementation, which looks nicer, but then require your current implementation to only be deimos specific if mask design information is present. Let's discuss on Slack or at the next telecon.

profxj commented on doc/deimos.rst:
```diff
@@ -57,7 +57,14 @@ Slit-mask design matching
 implemented only for DEIMOS and is switched on by setting **use_maskdesign** flag in
 :ref:`pypeit_par:EdgeTracePar Keywords` to *True*.  This is, already, the default for DEIMOS,
-except when *LongMirr* mask is used.
+except when *LongMirr* and *LVM* mask is used.
```
`and` -> `or the`

debora-pe replied:
Corrected

profxj commented on doc/dev/deimos_add_missing_obj.rst:
```diff
@@ -0,0 +1,132 @@
+- **force_fwhm**, which sets the initial FWHM in arcsec used for the optimal extraction of the spectra. The
+  default is 0.5 arcsec;
+
+- **mask_median_off**, which set a user provided offset in pixels between the measured and expected
```
typo `set` -> `sets`

debora-pe replied:
Corrected

profxj commented on doc/dev/deimos_radec_object.rst:
```diff
@@ -32,24 +33,29 @@ design information stored in the :class:`~pypeit.slittrace.SlitTraceSet` datamod
-then in arcsec) between the traced object and the center of the slit and, using the coordinates of the
-slit center available from the slitmask design, estimates the coordinates of the extracted spectrum.
+The slitmask design provides information on the distance of each targeted object from the slits edges,
+allowing to compares this expected distance to the one measured from the image. Differences between
```
the wording is a bit off here.

debora-pe replied:
Changed.

profxj commented on doc/dev/deimos_radec_object.rst:
```diff
@@ -32,24 +33,29 @@ design information stored in the :class:`~pypeit.slittrace.SlitTraceSet` datamod
-slit center available from the slitmask design, estimates the coordinates of the extracted spectrum.
+The slitmask design provides information on the distance of each targeted object from the slits edges,
+allowing to compares this expected distance to the one measured from the image. Differences between
+the expected and the measure slit length are also taken into account.
```
measure -> measured

debora-pe replied:
Corrected

profxj commented on doc/dev/deimos_radec_object.rst:
```diff
@@ -32,24 +33,29 @@ design information stored in the :class:`~pypeit.slittrace.SlitTraceSet` datamod
+the *Application* section of :ref:`deimos_slitmask_ids_report` must be set. Moreover, the **assign_obj** flag in
 :ref:`pypeit_par:SlitMaskPar Keywords` must be **True**.  This is the default for DEIMOS,
-except when *LongMirr* mask is used. One other keyword controls this procedure and it is **obj_toler**.
+except when *LongMirr* and *LVM* mask is used. One other keyword controls this procedure and it is **obj_toler**.
```
`and` -> `or the`

debora-pe replied:
Corrected

profxj reviewed (changes requested):
Nothing too substantial from me other than
an additional viewpoint here and there.

Great stuff!

jhennawi commented:
@deborape and @profxj, 

Okay I think I see the point. Namely that you are not able to assign objects from the mask until all the objects on all slits have been automatically identified by objfind, correct? So there is no way to make the control flow look the same for manual and mask assignment within the call to objfind, since you need to run objfind on all slits first. 

A reasonable compromise would then to leave the control flow as is, but add a symmetric set of data model flags for manual and force_extract objects, and to name the cards more informatively than force_extract which sounds like both. I don't think that needs to be in this PR, but we could at least add an issue on it so we don't forget.

debora-pe committed:
Merge branch 'develop' into deimos_force_extract

debora-pe committed:
changes to parsets

debora-pe committed:
fix docs

debora-pe committed:
change specobj attribute name

debora-pe committed:
added MANUAL_EXTRACT flag to SpecObj

debora-pe committed:
small fix to docs

debora-pe committed:
updated pypeit/tests/files

debora-pe committed:
updated CHANGES.rst

debora-pe committed:
updated pypeit/tests/files 2

debora-pe commented:
@profxj , @jhennawi 
I have addressed your comments on this PR. It is ready for you to review them. If it is OK with you, I will wait for feedback from Marla and then merge. But let me know if you have preferences on merging order with respect to `mosfire_oct20`.

The failing CI tests are probably due to the fact that I have created new cooked, but I have not uploaded them in the Dev Suite yet.
I have re-run the Dev Suite  and passed:
```
--- PYPEIT DEVELOPMENT SUITE PASSED 102/102 TESTS (Masters ignored) ---
Testing Started at 2021-04-14T21:21:10.595789
Testing Completed at 2021-04-15T03:31:37.819459
Total Time: 6:10:27.223670
```


And also the unit tests:
 ```
======================== test session starts =========================
platform darwin -- Python 3.8.5, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /Users/dpelliccia/PypeIt/PypeIt, configfile: setup.cfg
collected 259 items                                                  

test_alignments.py .                                           [  0%]
test_arc.py .                                                  [  0%]
test_arcimage.py ..                                            [  1%]
test_biasframe.py ..                                           [  2%]
test_bitmask.py ....                                           [  3%]
test_bpmimage.py ..                                            [  4%]
test_bspline.py ..........                                     [  8%]
test_buildimage.py .                                           [  8%]
test_calibrations.py .......                                   [ 11%]
test_chk_calibs.py ..                                          [ 12%]
test_coadd.py ..                                               [ 13%]
test_cooked.py .                                               [ 13%]
test_coremeta.py .                                             [ 13%]
test_datacontainer.py ........                                 [ 16%]
test_detector.py ...                                           [ 18%]
test_display.py .                                              [ 18%]
test_edgetrace.py .                                            [ 18%]
test_fitting.py ....                                           [ 20%]
test_flatfield.py .                                            [ 20%]
test_flexure.py .                                              [ 21%]
test_flux.py ....                                              [ 22%]
test_fluxspec.py ..                                            [ 23%]
test_frametype.py .                                            [ 23%]
test_load.py .                                                 [ 24%]
test_load_images.py .............                              [ 29%]
test_maskdesign_matching.py ...                                [ 30%]
test_masterframe.py .                                          [ 30%]
test_match.py ....                                             [ 32%]
test_metadata.py ......                                        [ 34%]
test_moment.py ......                                          [ 37%]
test_msgs.py ..                                                [ 37%]
test_opticalmodel.py ....                                      [ 39%]
test_parse.py ...                                              [ 40%]
test_pca.py ........                                           [ 43%]
test_processimage.py ..                                        [ 44%]
test_procimg.py .                                              [ 44%]
test_pydl.py .                                                 [ 45%]
test_pypeit.py ..                                              [ 45%]
test_pypeitimage.py .                                          [ 46%]
test_pypeitpar.py ................................             [ 58%]
test_pypeitsetup.py .......                                    [ 61%]
test_qa.py .                                                   [ 61%]
test_runpypeit.py ..                                           [ 62%]
test_scienceimage.py ...                                       [ 63%]
test_scripts.py .............                                  [ 68%]
test_setups.py .....................                           [ 76%]
test_skysub.py ..                                              [ 77%]
test_slitmask.py ...                                           [ 78%]
test_slittrace.py ....                                         [ 80%]
test_spec2dobj.py ......                                       [ 82%]
test_specobj.py ......                                         [ 84%]
test_specobjs.py .....                                         [ 86%]
test_spectrographs.py ....................                     [ 94%]
test_syncspec.py .                                             [ 94%]
test_traceimage.py .                                           [ 95%]
test_utils.py ...                                              [ 96%]
test_vcorr.py ..                                               [ 97%]
test_wavemodel.py .                                            [ 97%]
test_wavetilts.py ....                                         [ 99%]
test_wvcalib.py ..                                             [100%]

================== 259 passed in 940.60s (0:15:40) ===================
```

profxj reviewed (approved):
I am approving and I agree it is good to let
Marla 'kick the tires' before we merge.

profxj commented on pypeit/tests/test_slitmask.py:
```diff
@@ -77,6 +82,71 @@ def test_assign_maskinfo():
     os.remove(data_path('tst_sobjs.fits'))
 
 
+def test_add_missing_obj():
```
decorate this test with @cooked_required

debora-pe committed:
fix to tests/test_slitmask.py

debora-pe committed:
fix chk_version in show_2dspec

profxj committed:
vette me

jhennawi committed:
Update extract.py

jhennawi committed:
Update skysub.py

jhennawi commented on pypeit/reduce.py:
```diff
@@ -389,9 +389,20 @@ def run(self, ra=None, dec=None, obstime=None, std_trace=None, show_peaks=False,
-        if self.nobj > 0 and self.par['reduce']['slitmask']['assign_obj'] and self.slits.maskdef_designtab is not None:
-            self.slits.assign_maskinfo(self.sobjs_obj, self.get_platescale(None),
-                                       TOLER=self.par['reduce']['slitmask']['obj_toler'])
+        if self.par['reduce']['slitmask']['assign_obj'] and self.slits.maskdef_designtab is not None:
```
I somehow missed this on past PRs where this was added (i.e. assign_maskinfo). None of this should be here. This run method for reduce is abstracted to be independent of whether you are reducing Echelle, MultisSlit or IFU data. Likewise for the find_objects method. The find_object_pypeline is then a dummy method that gets overloaded by the instrument specific children. The same philosophy is used for the local_skysub_extract calls, i.e. the mode specific routines are part of the children. 

If you add Multislit specific function calls here with if then statements,  you are violating the philosophy that this abstraction is based on. It is equivalent to have if this spectrograph do this calls all throughout pypeit which we are committed to avoid. 

My suggestion is that these MultSlit specific function calls be added to find_objects_pypeline in the MultiSlit child, which is where anything that is multisit specific needs to be. I can help you do that, with relatively minor changes.

jhennawi reviewed (changes requested):
Hi @deborape, 

There is a flaw in how the  "assign_maskinfo" is being implemented. I think you didn't make this change but I somehow missed it when it was added. There should not be any mode specific calls in the parent class of reduce. See my detailed comments, but this needs to be fixed before we can merge.

jhennawi commented on pypeit/scripts/show_1dspec.py:
```diff
@@ -47,6 +47,10 @@ def main(args):
                     sobjs[ii].RA,
                     sobjs[ii].DEC,
                     sobjs[ii].MASKDEF_OBJNAME)
+            if sobjs[ii].MASKDEF_EXTRACT is not None and sobjs[ii].MASKDEF_EXTRACT is True:
```
Nice and thanks!

jhennawi reviewed (changes requested):
Hi @deborape, 

Thanks for addressing all of my comments. There is unfortunately something I caught in this PR that never should have been added to reduce. This was added in a previous PR and I somehow missed it, but now you are adding more multisit specific code in a place where there should be no mode specific code. Let's discuss. I think it is an easy fix, but these lines need to be removed and moved to their appropriate place before I can approve.

debora-pe committed:
added keys to SpecObj

debora-pe committed:
fix a bug with MANUAL_EXTRACT flag

profxj committed:
Merge pull request #1169 from pypeit/deimos_force_extract_jxp

Vette me

debora-pe committed:
move slitmask assign+force to multislit

debora-pe committed:
Merge branch 'deimos_force_extract' of https://github.com/pypeit/PypeIt into deimos_force_extract

profxj committed:
a little polishing

debora-pe commented:
@profxj , @jhennawi  I have moved the code that deals with the slitmask information in `find_objects_pypeline` in `MultiSlitReduce` class. Everything works. I re-run the tests and everything passes.
I will wait for `mosfire_oct20` to clear and I will merge it in.

debora-pe commented on pypeit/specobj.py:
```diff
@@ -211,10 +211,8 @@ def _init_internals(self):
-        self.maskwidth = None
 
         # Hand
-        self.hand_extract_flag = False
```
I think this should stay, otherwise for non manual extracted objects, this flag would be `None` instead of `False`. Is it correct?

profxj replied:
good catch, @deborape -- I thought it was just set to `None`
Please put it back and sorry about that.

debora-pe replied:
Ok. No problem. I'll do that.

debora-pe committed:
small fix

debora-pe committed:
fix conflict show_2dspec and tests files

debora-pe committed:
fix conflict pypeitpar.py

debora-pe committed:
fix conflict extract.py

debora-pe committed:
merge develop

debora-pe committed:
small fix to pypeitpar.py

debora-pe committed:
small fix to doc string

debora-pe committed:
merge deimos_collate_1d

debora-pe committed:
Merge branch 'develop' into deimos_force_extract

debora-pe committed:
updated unit test files

debora-pe commented:
Ok! All tests pass.

Dev Suite
```
--- PYPEIT DEVELOPMENT SUITE PASSED 108/108 TESTS (Masters ignored) ---
Testing Started at 2021-04-22T13:33:19.493201
Testing Completed at 2021-04-22T18:59:00.824461
Total Time: 5:25:41.331260
```

Unit tests

```
============================ test session starts ============================
platform darwin -- Python 3.8.5, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /Users/dpelliccia/PypeIt/PypeIt, configfile: setup.cfg
collected 271 items                                                         

test_alignments.py .                                                  [  0%]
test_arc.py .                                                         [  0%]
test_arcimage.py ..                                                   [  1%]
test_biasframe.py ..                                                  [  2%]
test_bitmask.py ....                                                  [  3%]
test_bpmimage.py ..                                                   [  4%]
test_bspline.py ..........                                            [  8%]
test_buildimage.py .                                                  [  8%]
test_calibrations.py .......                                          [ 11%]
test_chk_calibs.py ..                                                 [ 11%]
test_coadd.py ..                                                      [ 12%]
test_collate_1d.py .......                                            [ 15%]
test_cooked.py .                                                      [ 15%]
test_coremeta.py .                                                    [ 15%]
test_datacontainer.py ........                                        [ 18%]
test_detector.py ...                                                  [ 19%]
test_display.py .                                                     [ 20%]
test_edgetrace.py .                                                   [ 20%]
test_fitting.py ....                                                  [ 22%]
test_flatfield.py .                                                   [ 22%]
test_flexure.py .                                                     [ 22%]
test_flux.py ....                                                     [ 24%]
test_fluxspec.py ..                                                   [ 25%]
test_frametype.py .                                                   [ 25%]
test_history.py ....                                                  [ 26%]
test_load.py .                                                        [ 27%]
test_load_images.py .............                                     [ 32%]
test_maskdesign_matching.py ...                                       [ 33%]
test_masterframe.py .                                                 [ 33%]
test_match.py ....                                                    [ 35%]
test_metadata.py ......                                               [ 37%]
test_moment.py ......                                                 [ 39%]
test_msgs.py ..                                                       [ 40%]
test_opticalmodel.py ....                                             [ 41%]
test_parse.py ...                                                     [ 42%]
test_pca.py ........                                                  [ 45%]
test_processimage.py ..                                               [ 46%]
test_procimg.py .                                                     [ 46%]
test_pydl.py .                                                        [ 47%]
test_pypeit.py ..                                                     [ 47%]
test_pypeitimage.py .                                                 [ 48%]
test_pypeitpar.py ................................                    [ 60%]
test_pypeitsetup.py .......                                           [ 62%]
test_qa.py .                                                          [ 63%]
test_runpypeit.py ..                                                  [ 63%]
test_scienceimage.py ...                                              [ 64%]
test_scripts.py ..............                                        [ 70%]
test_setups.py .....................                                  [ 77%]
test_skysub.py ..                                                     [ 78%]
test_slitmask.py ...                                                  [ 79%]
test_slittrace.py ....                                                [ 81%]
test_spec2dobj.py ......                                              [ 83%]
test_specobj.py ......                                                [ 85%]
test_specobjs.py .....                                                [ 87%]
test_spectrographs.py ....................                            [ 94%]
test_syncspec.py .                                                    [ 95%]
test_traceimage.py .                                                  [ 95%]
test_utils.py ...                                                     [ 96%]
test_vcorr.py ..                                                      [ 97%]
test_wavemodel.py .                                                   [ 97%]
test_wavetilts.py ....                                                [ 99%]
test_wvcalib.py ..                                                    [100%]

====================== 271 passed in 948.63s (0:15:48) ======================
```

jhennawi reviewed (approved):
Great work! Thanks!!
